src/Simulator/IncrementalRedstoneSimulator/IncrementalRedstoneSimulator.cpp

```diff
@@ -0,0 +1,170 @@
+
+		case E_BLOCK_REDSTONE_TORCH_OFF:
+		case E_BLOCK_REDSTONE_TORCH_ON: return cpp14::make_unique<cRedstoneTorchHandler>(a_World);
+
+		case E_BLOCK_PISTON:
+		case E_BLOCK_STICKY_PISTON: return cpp14::make_unique<cPistonHandler>(a_World);
+
+		case E_BLOCK_LEVER:
+		case E_BLOCK_STONE_BUTTON:
+		case E_BLOCK_WOODEN_BUTTON: return cpp14::make_unique<cRedstoneToggleHandler>(a_World);
+
+		case E_BLOCK_BLOCK_OF_REDSTONE: return cpp14::make_unique<cRedstoneBlockHandler>(a_World);
+		case E_BLOCK_COMMAND_BLOCK: return cpp14::make_unique<cCommandBlockHandler>(a_World);
+		case E_BLOCK_NOTE_BLOCK: return cpp14::make_unique<cNoteBlockHandler>(a_World);
+		case E_BLOCK_REDSTONE_WIRE: return cpp14::make_unique<cRedstoneWireHandler>(a_World);
+		case E_BLOCK_TNT: return cpp14::make_unique<cTNTHandler>(a_World);
+		case E_BLOCK_TRAPPED_CHEST: return cpp14::make_unique<cTrappedChestHandler>(a_World);
+		case E_BLOCK_TRIPWIRE_HOOK: return cpp14::make_unique<cTripwireHookHandler>(a_World);
+		default:
+		{
+			if (cBlockDoorHandler::IsDoorBlockType(a_BlockType))
+			{
+				return cpp14::make_unique<cDoorHandler>(a_World);
+			}
+
+			if (cBlockInfo::FullyOccupiesVoxel(a_BlockType))
+			{
+				return cpp14::make_unique<cSolidBlockHandler>(a_World);
+			}
+			return nullptr;
+		}
+	}
+}
+
+
+
+
+
+void cIncrementalRedstoneSimulator::Simulate(float a_dt)
+{
+	for (auto & DelayInfo : m_Data.m_MechanismDelays)
+	{
+		if ((--DelayInfo.second.first) == 0)
+		{
+			m_Data.GetActiveBlocks().emplace_back(DelayInfo.first);
+		}
+	}
+
+	// Build our work queue
+	cVector3iArray WorkQueue;
+	std::swap(WorkQueue, m_Data.GetActiveBlocks());
+
+	// Process the work queue
+	while (!WorkQueue.empty())
+	{
+		// Grab the first element and remove it from the list
+		Vector3i CurrentLocation = WorkQueue.back();
+		WorkQueue.pop_back();
+
+		BLOCKTYPE CurrentBlock;
+		NIBBLETYPE CurrentMeta;
+		if (!m_World.GetBlockTypeMeta(CurrentLocation.x, CurrentLocation.y, CurrentLocation.z, CurrentBlock, CurrentMeta))
+		{
+			continue;
+		}
+
+		auto CurrentHandler = std::move(cIncrementalRedstoneSimulator::CreateComponent(m_World, CurrentBlock, &m_Data));
+		if (CurrentHandler == nullptr)
+		{
+			continue;
+		}
+
+		cRedstoneHandler::PoweringData Power;
+		for (const auto & Location : CurrentHandler->GetValidSourcePositions(CurrentLocation, CurrentBlock, CurrentMeta))
+		{
+			BLOCKTYPE PotentialBlock;
+			NIBBLETYPE PotentialMeta;
+			if ((Location.y < 0) || (Location.y > cChunkDef::Height))
+			{
+				continue;
+			}
+			m_World.GetBlockTypeMeta(Location.x, Location.y, Location.z, PotentialBlock, PotentialMeta);
+
+			auto PotentialSourceHandler = std::move(cIncrementalRedstoneSimulator::CreateComponent(m_World, PotentialBlock, &m_Data));
+			if (PotentialSourceHandler == nullptr)
+			{
+				continue;
+			}
+
+			decltype(Power) PotentialPower(PotentialBlock, PotentialSourceHandler->GetPowerDeliveredToPosition(Location, PotentialBlock, PotentialMeta, CurrentLocation, CurrentBlock));
```

Comment:
It might be worth adding a manual devirtualisation here. Yes it would be ugly but it might give a significant performace boost.

Something along the lines of:
```
auto Delivered = IsRedstone(PotentialBlock) 
    ? PotentialSourceHandler->GetPowerDeliveredToPosition(Location, PotentialBlock, PotentialMeta, CurrentLocation, CurrentBlock) 
    : PotentialSourceHandler->SolidBlockHandler::GetPowerDeliveredToPosition(Location, PotentialBlock, PotentialMeta, CurrentLocation, CurrentBlock) ;
decltype(Power) PotentialPower(PotentialBlock, Delivered);
```